### DISCUSSION TITLE
Is it possible to remove broker totally when Zeek working on standalone mode ?

### DISCUSSION THREAD

**Discussion by cxdinter** [zeek/zeek#2063] [General]
Hi,
   Maybe this question is stupid. But, I really want to know the work mechanism of broker. 
   According to the document, broker is used for the communication between different workers. So, if Zeek is running with standalone mode (only one Zeek process, configured by node.cfg), is there any chance to remove libbroker.so from Zeek install package ? (Concurrently, remove any broker API interface in Zeek source code (sure, huge work). )

   Thank you in advance.

**Comment by rsmmr** [comment]
No, not really. Broker's functionality is generally available/useful in standalone mode, too (e.g., data stores; receiving input from external clients).

At the code level, usage of Broker's API is baked into a quite a number of places, and they'd all have to be `#ifdef`ed out, plus some stubs to take their place and report errors in case it's user-visible (like Broker functions called from scrips). Would be hard to maintain a separate code path for that.

What's the concern with having `libbroker.so`?

**Reply by cxdinter** [reply]
> No, not really. Broker's functionality is generally available/useful in standalone mode, too (e.g., data stores; receiving input from external clients).
> 
> At the code level, usage of Broker's API is baked into a quite a number of places, and they'd all have to be `#ifdef`ed out, plus some stubs to take their place and report errors in case it's user-visible (like Broker functions called from scrips). Would be hard to maintain a separate code path for that.
> 
> What's the concern with having `libbroker.so`?

Hi, thanks for your quick response.

I am trying to deploy Zeek into embedded environment (iMX8 chip, running with Linux). The EMMC flash and DDR memory is very limited. The total Zeek install package is around 700MB, and running RAM is around 230MB. I try to optimize the package size, and already tried all disable compile arguments and removed some unused protocols in source code. But the total package still around 550MB. I see lib broker .so already occupy 210MB flash. So, I think if it is possible to totally remove it.

Any suggestion for this situation ? Thank you. 
Just one question, if I removed lib broker.so and cut all related dependencies in Zeek source code. The Zeek's behavior will be like what ? Does it still can monitor network and record logs ? (off course, by stand alone mode)

**Reply by Neverlord** [reply]
Did you try to build everything statically (by setting `BUILD_SHARED_LIBS` in CMake)? That should at least prune all symbols that Zeek doesn't actually use.

**Reply by cxdinter** [reply]
> BUILD_SHARED_LIBS

Hi, I think this is already set with true by default in Configure file. Right ?

**Reply by Neverlord** [reply]
Oh, Broker still has some quirky setup that doesn't honor that flag. Sorry! Broker has it's own custom flag called `ENABLE_STATIC_ONLY`.

Seems like the configure script has an option for that called `--enable-static-broker`: https://github.com/zeek/zeek/blob/master/configure#L65. You should get a Zeek version that doesn't link to `libbroker.so` (it may still get compiled, depending on how that configure option works).

**Reply by cxdinter** [reply]
> Oh, Broker still has some quirky setup that doesn't honor that flag. Sorry! Broker has it's own custom flag called `ENABLE_STATIC_ONLY`.
> 
> Seems like the configure script has an option for that called `--enable-static-broker`: https://github.com/zeek/zeek/blob/master/configure#L65. You should get a Zeek version that doesn't link to `libbroker.so` (it may still get compiled, depending on how that configure option works).

Thank you !
How about the size of static broker ? If it is much less than libbroker.so, I will try.

**Reply by JustinAzoff** [reply]
>  But the total package still around 550MB. I see lib broker .so already occupy 210MB flash.

Are you doing a non stripped debug build? See the help for `configure`:

```
--build-type=TYPE           set CMake build type [RelWithDebInfo]:
                                  - Debug: optimizations off, debug symbols + flags
                                  - MinSizeRel: size optimizations, debugging off
                                  - Release: optimizations on, debugging off
                                  - RelWithDebInfo: optimizations on,
                                        debug symbols on, debug flags off
```

you're probably using the default of RelWithDebInfo when you want to be using Release

**Reply by cxdinter** [reply]
> MinSizeRel

Hi, I really not use it. I will try 'MinSizeRe'. 
Thank you very much !

**Reply by cxdinter** [reply]
> > But the total package still around 550MB. I see lib broker .so already occupy 210MB flash.
> 
> Are you doing a non stripped debug build? See the help for `configure`:
> 
> ```
> --build-type=TYPE           set CMake build type [RelWithDebInfo]:
>                                   - Debug: optimizations off, debug symbols + flags
>                                   - MinSizeRel: size optimizations, debugging off
>                                   - Release: optimizations on, debugging off
>                                   - RelWithDebInfo: optimizations on,
>                                         debug symbols on, debug flags off
> ```
> 
> you're probably using the default of RelWithDebInfo when you want to be using Release

Hi Justin,
     After compiled Zeek with '--build-type=MinSizeRel', the total package size reduced to around 30MB. At that moment, I cannot trust my eyes :) 
     I am very appreciate !!! Thanks a lot.

**Reply by JustinAzoff** [reply]
Hi, you should definitely try --build-type=Release as well, MinSizeRel might only be a little smaller, but could be a lot slower.

**Reply by cxdinter** [reply]
> Hi, you should definitely try --build-type=Release as well, MinSizeRel might only be a little smaller, but could be a lot slower.

Hi, You are right. use '--build-type=Release' only increase 3MB size. Thanks !!
By the way, Zeek occupy about 220MB RAM during run time. If there any opportunity to reduce the RAM usage ? I tried to remove some scripts from local.zeek, but it is useless.
